### DISCUSSION TITLE
Raw socket support

### DISCUSSION THREAD

**Discussion by vaguue** [libuv/libuv#4251] [General]
I'm currently developing a Node.js application where I need support for raw sockets across various operating systems. I'm referencing libuv's code for this purpose. This has led me to wonder why libuv doesn't natively support raw sockets. Are there specific limitations or challenges that prevent this functionality from being implemented in libuv? Any insights would be greatly appreciated!

Thank you,
Seva D.

**Comment by bnoordhuis** [comment]
Platform-specific raw socket APIs are different enough that it's difficult to efficiently abstract over, often require elevated privileges, and aren't used much outside of ping- or traceroute-like tools.

On the whole it doesn't seem that useful to support in libuv. Cost/benefit tradeoff, basically.

**Reply by vaguue** [reply]
Got you, thanks

**Reply by vaguue** [reply]
Do I need to call functions to switch a created socket to non-blocking mode in the following manner:
```cpp
#ifdef _WIN32
  unsigned long flag = 1;
  if (ioctlsocket(pollfd, FIONBIO, &flag) != 0) {
    Napi::Error::New(env, "Error setting socket to non-blocking mode: " + getSystemError()).ThrowAsJavaScriptException();
    return;
  }
#else
  int flag = 1;
  if ((flag = fcntl(pollfd, F_GETFL, 0)) == -1) {
    Napi::Error::New(env, "Error setting socket to non-blocking mode: " + getSystemError()).ThrowAsJavaScriptException();
    return;
  }
  if (fcntl(pollfd, F_SETFL, flag | O_NONBLOCK) == -1) {
    Napi::Error::New(env, "Error setting socket to non-blocking mode: " + getSystemError()).ThrowAsJavaScriptException();
    return;
  }
#endif
```
Or is it sufficient to simply make a call to ```socket(domain, type, protocol)``` and pass it to ```uv_poll_init_socket```?
I've looked into the source code and noticed that basically this calls aren't necessary because the socket already gets switched to non-blocking mode by ```uv_poll_init_socket```, but I've also seen some projects where author calls this functions manually, so I decided to double-check by asking. 
@bnoordhuis

**Reply by bnoordhuis** [reply]
Libuv takes care of switching it to non-blocking mode.
